File: clients/ios/SsdidDrive/SsdidDrive/Data/Repository/RecoveryRepositoryImpl.swift
import Foundation

/// Implementation of RecoveryRepository using the API client
final class RecoveryRepositoryImpl: RecoveryRepository {

    // MARK: - Properties

    private let apiClient: APIClient

    // MARK: - Initialization

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    // MARK: - RecoveryRepository

    func setupRecovery(serverShare: String, keyProof: String) async throws {
        struct Request: Encodable {
            let server_share: String
            let key_proof: String
        }
        let _: EmptyResponse = try await apiClient.request(
            "/recovery/setup",
            method: .post,
            body: Request(server_share: serverShare, key_proof: keyProof),
            requiresAuth: true
        )
    }

    func getStatus() async throws -> RecoveryStatusResponse {
        return try await apiClient.request(
            "/recovery/status",
            method: .get,
            body: nil as String?,
            requiresAuth: true
        )
    }

    func getServerShare(did: String) async throws -> ServerShareResponse {
        let queryItems = [URLQueryItem(name: "did", value: did)]
        return try await apiClient.request(
            "/recovery/share",
            method: .get,
            body: nil as String?,
            queryItems: queryItems,
            requiresAuth: false
        )
    }

    func completeRecovery(
        oldDid: String,
        newDid: String,
        keyProof: String,
        kemPublicKey: String
    ) async throws -> CompleteRecoveryResponse {
        struct Request: Encodable {
            let old_did: String
            let new_did: String
            let key_proof: String
            let kem_public_key: String
        }
        return try await apiClient.request(
            "/recovery/complete",
            method: .post,
            body: Request(old_did: oldDid, new_did: newDid, key_proof: keyProof, kem_public_key: kemPublicKey),
            requiresAuth: false
        )
    }

    func deleteSetup() async throws {
        let _: EmptyResponse = try await apiClient.request(
            "/recovery/setup",
            method: .delete,
            body: nil as String?,
            requiresAuth: true
        )
    }

    // MARK: - Trustee Recovery

    func setupTrustees(threshold: Int, shares: [TrusteeShareRequest]) async throws {
        struct Request: Encodable {
            let threshold: Int
            let shares: [TrusteeShareRequest]
        }
        try await apiClient.requestNoContent(
            "/recovery/trustees/setup",
            method: .post,
            body: Request(threshold: threshold, shares: shares),
            requiresAuth: true
        )
    }

    func getTrustees() async throws -> [Trustee] {
        struct TrusteesResponse: Decodable {
            let trustees: [Trustee]
            let threshold: Int
        }
        let response: TrusteesResponse = try await apiClient.request(
            "/recovery/trustees",
            method: .get,
            body: nil as String?,
            requiresAuth: true
        )
        return response.trustees
    }

    func getPendingRequests() async throws -> [RecoveryRequest] {
        let response: PendingRequestsResponse = try await apiClient.request(
            "/recovery/requests/pending",
            method: .get,
            body: nil as String?,
            requiresAuth: true
        )
        return response.requests
    }

    func getHeldShares() async throws -> [RecoveryShare] {
        // Trustees see pending requests via getPendingRequests().
        // There is no separate held-shares endpoint — return empty.
        return []
    }

    func approveRequest(requestId: String) async throws {
        try await apiClient.requestNoContent(
            "/recovery/requests/\(requestId)/approve",
            method: .post,
            body: nil as String?,
            requiresAuth: true
        )
    }

    func rejectRequest(requestId: String) async throws {
        try await apiClient.requestNoContent(
            "/recovery/requests/\(requestId)/reject",
            method: .post,
            body: nil as String?,
            requiresAuth: true
        )
    }

    func getMyRecoveryRequest() async throws -> RecoveryRequest? {
        // The requester's own request is tracked client-side after createRecoveryRequest().
        // There is no "my request" endpoint — return nil.
        return nil
    }

    func createRecoveryRequest(did: String) async throws -> RecoveryRequest {
        struct Request: Encodable { let did: String }
        struct Response: Decodable {
            let id: String
            let did: String
            let status: String
            let approvedShares: Int
            let requiredShares: Int
            let expiresAt: Date
            let createdAt: Date

            enum CodingKeys: String, CodingKey {
                case id
                case did
                case status
                case approvedShares = "approved_shares"
                case requiredShares = "required_shares"
                case expiresAt = "expires_at"
                case createdAt = "created_at"
            }
        }
        let response: Response = try await apiClient.request(
            "/recovery/requests",
            method: .post,
            body: Request(did: did),
            requiresAuth: false
        )
        return RecoveryRequest(
            id: response.id,
            requesterId: response.did,
            requesterEmail: "",
            requesterName: nil,
            status: RecoveryRequest.Status(rawValue: response.status) ?? .pending,
            approvedShares: response.approvedShares,
            requiredShares: response.requiredShares,
            expiresAt: response.expiresAt,
            createdAt: response.createdAt
        )
    }

    func getReleasedShares(requestId: String, did: String) async throws -> [ReleasedShare] {
        struct Response: Decodable {
            let shares: [ReleasedShare]
        }
        let queryItems = [URLQueryItem(name: "did", value: did)]
        let response: Response = try await apiClient.request(
            "/recovery/requests/\(requestId)/shares",
            method: .get,
            body: nil as String?,
            queryItems: queryItems,
            requiresAuth: false
        )
        return response.shares
    }

    func initiateRecovery() async throws -> RecoveryRequest {
        // This endpoint is for the unauthenticated recovery flow (locked-out user).
        // From the settings/authenticated context there is no direct initiate endpoint.
        throw APIClient.APIError.serverError
    }
}

/// Empty response for endpoints that return 201/204
private struct EmptyResponse: Decodable {}

File: clients/android/app/src/main/kotlin/my/ssdid/drive/data/repository/RecoveryRepositoryImpl.kt
package my.ssdid.drive.data.repository

import my.ssdid.drive.data.remote.ApiService
import my.ssdid.drive.data.remote.dto.ApproveRequestResponse
import my.ssdid.drive.data.remote.dto.CompleteRecoveryRequest
import my.ssdid.drive.data.remote.dto.CompleteRecoveryResponse
import my.ssdid.drive.data.remote.dto.CreateRecoveryRequestBody
import my.ssdid.drive.data.remote.dto.ListTrusteesResponse
import my.ssdid.drive.data.remote.dto.MyRecoveryRequestResponse
import my.ssdid.drive.data.remote.dto.PendingRequestsResponse
import my.ssdid.drive.data.remote.dto.RecoveryRequestResponse
import my.ssdid.drive.data.remote.dto.RecoveryStatusResponse
import my.ssdid.drive.data.remote.dto.RejectRequestResponse
import my.ssdid.drive.data.remote.dto.ReleasedSharesResponse
import my.ssdid.drive.data.remote.dto.ServerShareResponse
import my.ssdid.drive.data.remote.dto.SetupRecoveryRequest
import my.ssdid.drive.data.remote.dto.SetupTrusteesRequest
import my.ssdid.drive.data.remote.dto.SetupTrusteesResponse
import my.ssdid.drive.domain.repository.RecoveryRepository
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RecoveryRepositoryImpl @Inject constructor(
    private val apiService: ApiService
) : RecoveryRepository {

    override suspend fun setupRecovery(serverShare: String, keyProof: String): Result<Unit> = try {
        val response = apiService.setupRecovery(SetupRecoveryRequest(serverShare, keyProof))
        if (response.isSuccessful) Result.success(Unit)
        else Result.failure(Exception("Setup failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun getStatus(): Result<RecoveryStatusResponse> = try {
        val response = apiService.getRecoveryStatus()
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Status check failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun getServerShare(did: String): Result<ServerShareResponse> = try {
        val response = apiService.getRecoveryShare(did)
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Share retrieval failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun completeRecovery(
        oldDid: String, newDid: String, keyProof: String, kemPublicKey: String
    ): Result<CompleteRecoveryResponse> = try {
        val response = apiService.completeRecovery(
            CompleteRecoveryRequest(oldDid, newDid, keyProof, kemPublicKey)
        )
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Recovery failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun deleteSetup(): Result<Unit> = try {
        val response = apiService.deleteRecoverySetup()
        if (response.isSuccessful) Result.success(Unit)
        else Result.failure(Exception("Delete failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    // ==================== Trustee Management ====================

    override suspend fun setupTrustees(request: SetupTrusteesRequest): Result<SetupTrusteesResponse> = try {
        val response = apiService.setupTrustees(request)
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Trustee setup failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun getTrustees(): Result<ListTrusteesResponse> = try {
        val response = apiService.getTrustees()
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Get trustees failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    // ==================== Recovery Requests — Authenticated ====================

    override suspend fun initiateRecoveryRequest(): Result<RecoveryRequestResponse> = try {
        val response = apiService.initiateRecoveryRequest()
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Initiate recovery failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun getMyRecoveryRequest(): Result<MyRecoveryRequestResponse> = try {
        val response = apiService.getMyRecoveryRequest()
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Get my recovery request failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun getPendingRecoveryRequests(): Result<PendingRequestsResponse> = try {
        val response = apiService.getPendingRecoveryRequests()
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Get pending requests failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun approveRecoveryRequest(requestId: String): Result<ApproveRequestResponse> = try {
        val response = apiService.approveRecoveryRequest(requestId)
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Approve request failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun rejectRecoveryRequest(requestId: String): Result<RejectRequestResponse> = try {
        val response = apiService.rejectRecoveryRequest(requestId)
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Reject request failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    // ==================== Recovery Requests — Unauthenticated ====================

    override suspend fun createRecoveryRequest(did: String): Result<RecoveryRequestResponse> = try {
        val response = apiService.createRecoveryRequest(CreateRecoveryRequestBody(did))
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Create recovery request failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }

    override suspend fun getReleasedShares(requestId: String, did: String): Result<ReleasedSharesResponse> = try {
        val response = apiService.getReleasedShares(requestId, did)
        if (response.isSuccessful) Result.success(response.body()!!)
        else Result.failure(Exception("Get released shares failed: ${response.code()}"))
    } catch (e: Exception) {
        Result.failure(e)
    }
}
